src/reformulate.jl

```diff
@@ -1,82 +1,387 @@
+"""
 
-Reformulate disjunction.
 """
-function reformulate_disjunction(m::Model, disj...; bin_var, reformulation, param)
-    #check disj
-    disj = [check_constraint!(m, constr) for constr in disj]#check_disjunction!(m, disj)
-    #get original variable refs and variable names
-    vars = setdiff(all_variables(m), m[bin_var])
-    var_names = unique(Symbol.([split("$var","[")[1] for var in vars]))
-    if !in(:gdp_variable_refs, keys(object_dictionary(m)))
-        @expression(m, gdp_variable_refs, vars)
+function _reformulate_disjunctions(model::JuMP.Model, method::AbstractReformulationMethod)
+    for (disj_idx, disj) in _disjunctions(model)
+        _reformulate_disjuncts(model, disj_idx, disj, method)
     end
-    if !in(:gdp_variable_names, keys(object_dictionary(m)))
-        @expression(m, gdp_variable_names, var_names)
+end
+
+function _reformulate_disjuncts(model::JuMP.Model, disj_idx::DisjunctionIndex, disj::ConstraintData{T}, method::Union{BigM,Indicator}) where {T<:Disjunction}
+    for d in disj.constraint.disjuncts
+        _reformulate_disjunct(model, d, method)
     end
-    #run reformulation
-    if reformulation == :hull
-        disaggregate_variables(m, disj, bin_var)
-        sum_disaggregated_variables(m, disj, bin_var)
+end
+
+function _reformulate_disjuncts(model::JuMP.Model, disj_idx::DisjunctionIndex, disj::ConstraintData{T}, method::Hull) where {T<:Disjunction}
+    disj_vrefs = _get_disjunction_variables(disj)
+    _update_variable_bounds.(disj_vrefs) 
+    for d in disj.constraint.disjuncts #reformulate each disjunct
+        _disaggregate_variables(model, d, disj_vrefs) #disaggregate variables for that disjunct
+        _reformulate_disjunct(model, d, method)
+    end
+    for vref in disj_vrefs #create sum constraint for disaggregated variables
+        _aggregate_variable(model, vref, disj_idx)
     end
-    reformulate_disjunction(disj, bin_var, reformulation, param)
+end
 
-    #show new constraints as a Dict
-    new_constraints = Dict{Symbol,Any}(
-        Symbol(bin_var,"[$i]") => disj[i] for i in eachindex(disj)
-    )
-    new_constraints[Symbol(bin_var,"_XOR")] = constraint_by_name(m, "XOR(disj_$bin_var)")
-    if reformulation == :hull
-        for var in m[:gdp_variable_refs]
-            agg_con_name = "$(var)_$(bin_var)_aggregation"
-            new_constraints[Symbol(agg_con_name)] = constraint_by_name(m, agg_con_name)
-        end
+"""
+
+"""
+function _reformulate_disjunct(model::JuMP.Model, d::Disjunct, method::AbstractReformulationMethod)
+    #reformulate each constraint and add to the model
+    ind_idx = JuMP.index(d.indicator) #logical indicator
+    bv_idx = _indicator_to_binary(model)[ind_idx]
+    bvref = JuMP.VariableRef(model, bv_idx)
+    for con in d.constraints
+        cdata = _disjunct_constraints(model)[JuMP.index(con)]
+        _reformulate_disjunctive_constraint(model, cdata.constraint, bvref, method)
```

Comment:
Is this method needed or used?

Reply:
Yes, it is called from `_reformulate_disjuncts`. The reformulation is made on each disjunct of the disjunction.

Reply:
Right, but is this particular dispatch used with `AbstractReformulationMethod`? It seems you end up having to make a dispatch for each particular method.

Reply:
Yes you dispatch for each method at the disjunct constraint level